### DISCUSSION TITLE
SVS to DZI results in pink images

### DISCUSSION THREAD

**Discussion by Gautamji** [libvips/libvips#2899] [Q&A]
## Bug report

**Describe the bug**
I have been trying to use https://github.com/zoellner/sharp-heic-lambda-layer as it already has an implementation of a custom libvips with sharp.
I added openslide to it.
And am using it to convert openslide images to dzi.

When I convert an SVS image to DZI, it turns pink.

**To Reproduce**
This is the image I need to convert
https://drive.google.com/file/d/16G7jiIpMfdeBb5aXTrhE_43aQBUbRZgs/view?usp=sharing 

2. Use Configuration '....'
This is my slightly modified makefile from what was there in https://github.com/zoellner/sharp-heic-lambda-layer

https://gist.github.com/Gautamji/4b485d866672a8aab22de041ed6bdb8f

**Expected behavior**
I expect the image to convert normally

**Actual behavior**
It's turning pink

**Screenshots**
![Screenshot 2022-06-27 070714](https://user-images.githubusercontent.com/4068661/175844490-5cb1dff7-fc8d-44f2-af95-a5cc428fb828.png)

**Environment**
(please complete the following information)
 - OS: Linux (running on AWS Lambda)
 - Vips: 8.12.2

**Additional context**
I have been struggling with this for quite some time now, any help would be greatly appreciated.

**Comment by jcupitt** [comment]
Hi @Gautamji,

That's a weird one. I tried here:

```
$ time vips dzsave input.tiff x

real	0m56.717s
user	6m31.480s
sys	13m50.028s
```

And for `x_files/9/0_0.jpeg` I see:

![0_0](https://user-images.githubusercontent.com/580843/175934264-b5c49484-4e28-4d9c-9381-6310f9421a09.jpeg)

So I don't think it's a libvips bug -- I would guess something has gone wrong with your openslide install.

Does it work if you convert using the `vips` command line tool? What do you see for `vipsheader input.tiff`? I see:

```
$ vipsheader input.tiff 
input.tiff: 53784x32489 uchar, 4 bands, srgb, openslideload
```

**Reply by Gautamji** [reply]
Hi,
Thank you so much for such a quick reply. 
You are right, I don't think it is a vips issue.
It worked as it should have when I installed and used vips directly in a container.

I have been using the node sharp library for this and I think there may be some issue when I am trying to build it with a custom vips.

Is there some way I can input a byte stream directly from S3 into vips as I do in sharp? I am trying to see if I can push the container itself to lambda and run vips through the shell instead of using sharp.

**Reply by jcupitt** [reply]
Sure, there are some notes here: https://www.libvips.org/2019/12/11/What%27s-new-in-8.9.html

**Reply by Gautamji** [reply]
Hi,
I think I have found what could be the source of this issue.
Please let me know if there is a workaround for this or if I am using something wrong.

I had been using  custom built node sharp library with jpg2000 support for this initially when I first encountered this issue.
Thinking that it may be a sharp issue or that I may have built something wrong, I instead created a lambda docker image  and installed vips to it so that I could run it directly from the commandline.

this is the command I am using to convert the svs image 
`aws s3 cp s3://drbb-slides/input.svs - | vips --vips-progress dzsave stdin ./inputsvs-aws`

And this is resulting in a similar pink svs.

Then I ran it directly in the ubuntu terminal and it gave this additional warning 
`(vips:67091): VIPS-WARNING **: 11:41:31.162: IO error during reading of "ICC Profile"; tag ignored
`
And also resulted in a similarly pink svs

This svs has been set to public, so you should be able to run the command directly without needing credentials.

Please guide me on what I should next

**Reply by jcupitt** [reply]
I tried again, and it's still working for me. 

I couldn't download your s3 version (I get "unable to locate credentials"), but I downloaded the google drive version:

```
~/pics/openslide $ vipsheader input.svs
input.svs: 53784x32489 uchar, 4 bands, srgb, openslideload
~/pics/openslide $ md5sum input.svs
053025d9a66efb03251ded99859732c3  input.svs
~/pics/openslide $ vips dzsave input.svs x
```

Here's `x_files/9/0_0.jpeg`:

![0_0](https://user-images.githubusercontent.com/580843/180739832-3ceb6534-fe30-4ff2-9c6d-74a27264c2b4.jpeg)

What do you get for `vipsheader input.svs`?

**Reply by Gautamji** [reply]
Thank you for such a quick response,

I tried running
`aws s3 cp s3://drbb-slides/input.svs - | vipsheader stdin`

this was the result
`53784x32489 uchar, 3 bands, srgb, tiffload_source`

**Reply by jcupitt** [reply]
Ah, you've built libvips without openslide support. You have:

```
53784x32489 uchar, 3 bands, srgb, tiffload_source
```

You should have:

```
input.svs: 53784x32489 uchar, 4 bands, srgb, openslideload
```

SVS images are not true TIFFs, you need a special loader for them.

**Reply by Gautamji** [reply]
But I had built it with openslide support. 
this is what I get if I try with a local file instead of it being streamed via s3
`$ vipsheader ./input.svs`
`./input.svs: 53784x32489 uchar, 4 bands, srgb, openslideload`

**Reply by jcupitt** [reply]
Ahhh I see. No, openslide can't read from a pipe, it must have the slide in local storage. This is because formats like MRXS come in many separate files.

You'll need to do:

```
aws s3 cp s3://drbb-slides/input.svs /tmp
vips dzsave /tmp/input.svs something.zip
aws s3 cp something.zip s3://mybucket
```

And then unzip on S3 somehow.

**Reply by jcupitt** [reply]
with libvips 8.13, you can `dzsave` directly to a pipe, eg.:

```
vips dzsave_target /tmp/input.svs target[descriptor=1] | \
    aws s3 cp - s3://mybucket/something.zip
```

You'll still need to unzip on s3 though.
